toolchain/base/int_store.h

```diff
@@ -0,0 +1,422 @@
+    out << "int [";
+    if (is_value()) {
+      out << "value: " << AsValue() << "]";
+    } else if (is_index()) {
+      out << "index: " << AsIndex() << "]";
+    } else {
+      CARBON_CHECK(!is_valid());
+      out << "invalid]";
+    }
+  }
+
+  friend constexpr auto operator==(IntId lhs, IntId rhs) -> bool {
+    return lhs.id_ == rhs.id_;
+  }
+  friend constexpr auto operator<=>(IntId lhs, IntId rhs)
+      -> std::strong_ordering {
+    return lhs.id_ <=> rhs.id_;
+  }
+
+ private:
+  friend class IntStore;
+  friend Testing::IntStoreTestPeer;
+
+  // The shift needed when adjusting a between a `TokenIdBits`-width integer and
+  // a 32-bit integer.
+  static constexpr int TokenIdBitsShift = 32 - TokenIdBits;
+
+  // The maximum embedded value in an ID.
+  static constexpr int32_t MaxValue =
+      std::numeric_limits<int32_t>::max() >> TokenIdBitsShift;
+
+  // The ID value that represents an index of `0`. This is the first ID value
+  // representing an index, and all indices are `<=` to this.
+  //
+  // Because this is the first index ID, and we encoded indices as successive
+  // negative numbers counting downwards, we can both use a comparison with
+  // this ID to distinguish value and index IDs, and to compute the actual index
+  // from the ID. The computation of an index in fact is just a subtraction:
+  // `ZeroIndexId - id_`. Subtraction is *also* how most CPUs implement the
+  // comparison, and so all of this ends up carefully constructed to enable very
+  // small code size when testing for an embedded value and when that test fails
+  // computing and using the index.
+  static constexpr int32_t ZeroIndexId = std::numeric_limits<int32_t>::min() >>
+                                         (TokenIdBitsShift + 1);
+
+  // The minimum embedded value in an ID.
+  static constexpr int32_t MinValue = ZeroIndexId + 1;
+
+  // The invalid ID, which needs to be placed after the largest index, which
+  // count downwards as IDs so below the smallest index ID, in order to optimize
+  // the code sequence needed to distinguish between integer and value IDs and
+  // to convert index IDs into actual indices small.
+  static constexpr int32_t InvalidId = std::numeric_limits<int32_t>::min();
+
+  // The invalid index. This is the result of converting an invalid ID into an
+  // index. We ensure that conversion can be done so that we can simplify the
+  // code that first tries to use an embedded value, then converts to an index
+  // and checks that the index is valid.
+  static const int32_t InvalidIndex;
+
+  // Document the specific values of some of these constants to help visualize
+  // how the bit patterns map from the above computations.
+  //
+  // Each bit is either `T` for part of the token or `P` as part
+  // of the available payload that we use for the ID:
+  //
+  // clang-format off: visualizing bit positions
+  //
+  //                           0bTTTT'TTTT'TPPP'PPPP'PPPP'PPPP'PPPP'PPPP
```

Comment:
Nuanced thing, the `clang-format` throws me a little, maybe one of:

```suggestion
  // clang-format off: visualizing bit positions
  //
  // Each bit is either `T` for part of the token or `P` as part
  // of the available payload that we use for the ID:
  //
  //                           0bTTTT'TTTT'TPPP'PPPP'PPPP'PPPP'PPPP'PPPP
```

```suggestion
  // Each bit is tagged either `T` for part of the token or `P` as part
  // of the available payload that we use for the ID.
  //
  // clang-format off: visualizing bit positions
  //
  //                           0bTTTT'TTTT'TPPP'PPPP'PPPP'PPPP'PPPP'PPPP
```